src/librustc_mir/transform/const_prop.rs

```diff
@@ -504,36 +489,48 @@ impl<'mir, 'tcx> ConstPropagator<'mir, 'tcx> {
         }
     }
 
-    fn eval_place(&mut self, place: &Place<'tcx>, source_info: SourceInfo) -> Option<OpTy<'tcx>> {
+    fn eval_place(&mut self, place: &Place<'tcx>) -> Option<OpTy<'tcx>> {
         trace!("eval_place(place={:?})", place);
-        self.use_ecx(source_info, |this| this.ecx.eval_place_to_op(place, None))
+        self.use_ecx(|this| this.ecx.eval_place_to_op(place, None))
     }
 
     fn eval_operand(&mut self, op: &Operand<'tcx>, source_info: SourceInfo) -> Option<OpTy<'tcx>> {
         match *op {
             Operand::Constant(ref c) => self.eval_constant(c, source_info),
             Operand::Move(ref place) | Operand::Copy(ref place) => {
-                self.eval_place(place, source_info)
+                self.eval_place(place)
             }
         }
     }
 
-    fn check_unary_op(&mut self, arg: &Operand<'tcx>, source_info: SourceInfo) -> Option<()> {
-        self.use_ecx(source_info, |this| {
-            let ty = arg.ty(&this.local_decls, this.tcx);
-
-            if ty.is_integral() {
-                let arg = this.ecx.eval_operand(arg, None)?;
-                let prim = this.ecx.read_immediate(arg)?;
-                // Need to do overflow check here: For actual CTFE, MIR
-                // generation emits code that does this before calling the op.
-                if prim.to_bits()? == (1 << (prim.layout.size.bits() - 1)) {
-                    throw_panic!(OverflowNeg)
-                }
-            }
+    fn report_panic_as_lint(&self, source_info: SourceInfo, panic: PanicInfo<u64>) -> Option<()> {
+        // Somewhat convoluted way to re-use the CTFE error reporting code.
+        let lint_root = self.lint_root(source_info)?;
+        let error = InterpError::MachineStop(Box::new(format!("{:?}", panic)));
+        let mut diagnostic = error_to_const_error(&self.ecx, error.into());
+        diagnostic.span = source_info.span; // fix the span
+        diagnostic.report_as_lint(
+            self.tcx.at(source_info.span),
+            "this expression will panic at runtime",
+            lint_root,
+            None,
+        );
+        None
+    }
```

Comment:
So this is quite a mess. And "this expression will panic at runtime" is factually wrong; run-time panics are caused by `Assert` terminators which do not go through this code-path. But that confusion is pre-existing.

Probably, this should not use `ConstEvalErr` at all, but I have no idea which part of `ConstEvalErr::struct_generic` is geared specifically towards const-prop and should be put here (and removed in its original source).

Reply:
Probably the best thing would be to take the error reporting code from `visit_terminator`, put it into a helper function, and use that. That would also make things more consistent in terms of errors being reported on MIR with and without overflow checks.

I am holding off on this though because I fear that might create some dead code in `ConstEvalErr::struct_generic` -- was some code there added just to handle const-prop? If yes, that should be removed then.

Reply:
> And "this expression will panic at runtime" is factually wrong; run-time panics are caused by Assert terminators which do not go through this code-path. But that confusion is pre-existing.

right, it used to be true at the start. Maybe we could just go with a generic "this expression contains erroneous code"?

> Probably, this should not use ConstEvalErr at all, but I have no idea which part of ConstEvalErr::struct_generic is geared specifically towards const-prop and should be put here (and removed in its original source).

You can just report a lint directly via `self.tcx.sess.span_lint`. After that you should be able to make `MachineError` a `bug!` in `report_as_lint` again.

All this should probably be in separate PRs though, at least the lint and the message change are likely to be noisy.

Reply:
> I am holding off on this though because I fear that might create some dead code in ConstEvalErr::struct_generic -- was some code there added just to handle const-prop? If yes, that should be removed then.

There shouldn't be, we ignore everything but panic-likes in const prop

Reply:
> There shouldn't be, we ignore everything but panic-likes in const prop

I was referring to this mess:
https://github.com/rust-lang/rust/blob/4d1241f5158ffd66730e094d8f199ed654ed52ae/src/librustc/mir/interpret/error.rs#L146-L167

Is some of this const-prop hackery, or is this all still needed even if const-prop uses another code path?

Reply:
> After that you should be able to make MachineError a bug! in report_as_lint again.

No I don't think so, CTFE itself still [relies on this](https://github.com/rust-lang/rust/pull/68969/files#diff-388ebaf79f9ee3eb60b3ed7ea933fa66R23).

Reply:
Oh, and another question: does it really make sense for "const-eval happened and raised an error" to use the same lint as "const-prop did partial evaluation and determined that whenever this code runs, it will raise an error"? These seem like different things to me (and not being able to distinguish them has been quite confusing in my testing).

Reply:
Yea CTFE uses this in the `const_eval_raw` query.

> These seem like different things to me (and not being able to distinguish them has been quite confusing in my testing).

They are indeed quite different. The reason they are somewhat merged is that CTFE used to always emit a hard error, but we can't do that anymore. Maybe we should just make CTFE lints a future incompat lint (if we can get enough ppl behind that idea, if not just make it its own lint).

Reply:
> Yea CTFE uses this in the const_eval_raw query.

I'm aware it uses that function; my question is -- are all of the branches there still reachable, or is some of that stuff necessary just for const-prop? I have no idea what all this is about and why it is so messy and why `message` is ignored under some circumstances.

> They are indeed quite different. The reason they are somewhat merged is that CTFE used to always emit a hard error, but we can't do that anymore. Maybe we should just make CTFE lints a future incompat lint (if we can get enough ppl behind that idea, if not just make it its own lint).

I was more thinking of leaving CTFE with `const_err` and making const-prop emit some other error... that error is not necessarily about "consts" anyway. (I am also not sure if the const-prop errors really should be deny-by-default but that is a separate concern.)

Reply:
> I was more thinking of leaving CTFE with const_err and making const-prop emit some other error... that error is not necessarily about "consts" anyway.

the relevant rfc https://github.com/rust-lang/rfcs/pull/1229 mentions "erroneous code", though that's very vague. 

> (I am also not sure if the const-prop errors really should be deny-by-default but that is a separate concern.)

They are only deny-by-default because they share the same lint name, after splitting it making it warn-by-default seems ok. Also promoteds should share that lint.

Reply:
> Also promoteds should share that lint.

Aaaahhh, but then it'll become messy again because this means the `ConstEvalError` thing will also emit that lint.^^

Also, I think even ignoring that concern I only agree for promoteds that were promoted from a function, not those promoted from a `const`.

---

So, what about [this mess](https://github.com/rust-lang/rust/blob/4d1241f5158ffd66730e094d8f199ed654ed52ae/src/librustc/mir/interpret/error.rs#L146-L167), can anyone add comments to explain it and maybe remove what is not needed any more when const-prop switches away?

Reply:
> Also, I think even ignoring that concern I only agree for promoteds that were promoted from a function, not those promoted from a const.

yes that makes sense

> So, what about this mess, can anyone add comments to explain it and maybe remove what is not needed any more when const-prop switches away?

We can't change anything here as far as I can tell. This is all needed for CTFE. const prop just made use of the "lint only" case since it only ever fed panic errors into this function.

Reply:
Whatever "lint only case" means...

Reply:
I added a commit that (a) cleans up the control flow a bit and (b) adds some FIXMEs fur stuff that is weird and has no comment explaining it.